File: Src/java/engine/src/test/java/org/opencds/cqf/cql/engine/model/CachingModelResolverDecoratorTest.java
package org.opencds.cqf.cql.engine.model;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.isA;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.Date;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.opencds.cqf.cql.engine.data.CompositeDataProvider;
import org.opencds.cqf.cql.engine.retrieve.RetrieveProvider;

// TODO: Extend testing to cover more of the CachedModelResolver
class CachingModelResolverDecoratorTest {
    @Mock
    private ModelResolver mockModelResolver;

    @Mock
    private RetrieveProvider mockRetrieveProvider;

    private AutoCloseable mocks;

    @BeforeEach
    void before() {
        mocks = MockitoAnnotations.openMocks(this);
    }

    @AfterEach
    void after() throws Exception {
        mocks.close();
    }

    @Test
    @SuppressWarnings("deprecation")
    void context_path_resolved_only_once() {
        var m = mock(ModelResolver.class);
        when(m.getPackageName()).thenReturn("test.package");
        when(m.getContextPath("Patient", "Patient")).thenReturn("id");

        var cache = new CachingModelResolverDecorator(m);
        cache.getContextPath("Patient", "Patient");
        var result = cache.getContextPath("Patient", "Patient");

        assertEquals("id", result);
        verify(m, times(1)).getContextPath("Patient", "Patient");
    }

    @Test
    @SuppressWarnings({"unchecked", "deprecation"})
    void type_resolved_only_once() {
        var m = mock(ModelResolver.class);
        when(m.getPackageName()).thenReturn("test.package");
        when(m.resolveType(isA(Integer.class))).thenReturn((Class) Integer.class);
        when(m.resolveType(isA(Class.class))).thenThrow(new RuntimeException("Can't get a class of a class"));

        var cache = new CachingModelResolverDecorator(m);
        cache.resolveType(5);
        var result = cache.resolveType(5);

        assertEquals(Integer.class, result);
        verify(m, times(1)).resolveType(5);
    }

    @Test
    void resolveIdString() {
        final String object = "object";
        final String id = "text";

        when(mockModelResolver.resolveId(object)).thenReturn(id);

        final CompositeDataProvider compositeDataProvider =
                new CompositeDataProvider(mockModelResolver, mockRetrieveProvider);

        assertEquals(id, compositeDataProvider.resolveId(object));
        verify(mockModelResolver, times(1)).resolveId(object);
    }

    @Test
    void resolveIdIntLong() {
        final long object = 1L;
        final String id = "oneL";

        when(mockModelResolver.resolveId(object)).thenReturn(id);

        final CompositeDataProvider compositeDataProvider =
                new CompositeDataProvider(mockModelResolver, mockRetrieveProvider);

        assertEquals(id, compositeDataProvider.resolveId(object));
        verify(mockModelResolver, times(1)).resolveId(object);
    }

    @Test
    void resolveIdDate() {
        final Date object = new Date();
        final String id = "now";

        when(mockModelResolver.resolveId(object)).thenReturn(id);

        final CompositeDataProvider compositeDataProvider =
                new CompositeDataProvider(mockModelResolver, mockRetrieveProvider);

        assertEquals(id, compositeDataProvider.resolveId(object));
        verify(mockModelResolver, times(1)).resolveId(object);
    }
}

File: Src/java/engine/src/jvmTest/kotlin/org/opencds/cqf/cql/engine/model/CachingModelResolverDecoratorTest.kt
package org.opencds.cqf.cql.engine.model

import java.lang.AutoCloseable
import java.util.*
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.MockitoAnnotations
import org.opencds.cqf.cql.engine.data.CompositeDataProvider
import org.opencds.cqf.cql.engine.retrieve.RetrieveProvider

// TODO: Extend testing to cover more of the CachedModelResolver
internal class CachingModelResolverDecoratorTest {
    @Mock private val mockModelResolver: ModelResolver? = null

    @Mock private val mockRetrieveProvider: RetrieveProvider? = null

    private var mocks: AutoCloseable? = null

    @BeforeEach
    fun before() {
        mocks = MockitoAnnotations.openMocks(this)
    }

    @AfterEach
    @Throws(Exception::class)
    fun after() {
        mocks!!.close()
    }

    @Test
    @Suppress("deprecation")
    fun context_path_resolved_only_once() {
        val m = Mockito.mock(ModelResolver::class.java)
        Mockito.`when`(m.packageName).thenReturn("test.package")
        Mockito.`when`(m.getContextPath("Patient", "Patient")).thenReturn("id")

        val cache = CachingModelResolverDecorator(m)
        cache.getContextPath("Patient", "Patient")
        val result = cache.getContextPath("Patient", "Patient")

        Assertions.assertEquals("id", result)
        Mockito.verify(m, Mockito.times(1)).getContextPath("Patient", "Patient")
    }

    @Test
    @Suppress("deprecation")
    fun type_resolved_only_once() {
        val m = Mockito.mock(ModelResolver::class.java)
        Mockito.`when`(m.packageName).thenReturn("test.package")
        Mockito.`when`(m.resolveType(ArgumentMatchers.isA(Int::class.java)))
            .thenReturn(Int::class.java)
        Mockito.`when`(m.resolveType(ArgumentMatchers.isA(Class::class.java)))
            .thenThrow(RuntimeException("Can't get a class of a class"))

        val cache = CachingModelResolverDecorator(m)
        cache.resolveType(5)
        val result = cache.resolveType(5)

        Assertions.assertEquals(Int::class.java, result)
        Mockito.verify(m, Mockito.times(1)).resolveType(5)
    }

    @Test
    fun resolveIdString() {
        val `object` = "object"
        val id = "text"

        Mockito.`when`(mockModelResolver!!.resolveId(`object`)).thenReturn(id)

        val compositeDataProvider = CompositeDataProvider(mockModelResolver, mockRetrieveProvider)

        Assertions.assertEquals(id, compositeDataProvider.resolveId(`object`))
        Mockito.verify(mockModelResolver, Mockito.times(1)).resolveId(`object`)
    }

    @Test
    fun resolveIdIntLong() {
        val `object` = 1L
        val id = "oneL"

        Mockito.`when`(mockModelResolver!!.resolveId(`object`)).thenReturn(id)

        val compositeDataProvider = CompositeDataProvider(mockModelResolver, mockRetrieveProvider)

        Assertions.assertEquals(id, compositeDataProvider.resolveId(`object`))
        Mockito.verify(mockModelResolver, Mockito.times(1)).resolveId(`object`)
    }

    @Test
    fun resolveIdDate() {
        val `object` = Date()
        val id = "now"

        Mockito.`when`(mockModelResolver!!.resolveId(`object`)).thenReturn(id)

        val compositeDataProvider = CompositeDataProvider(mockModelResolver, mockRetrieveProvider)

        Assertions.assertEquals(id, compositeDataProvider.resolveId(`object`))
        Mockito.verify(mockModelResolver, Mockito.times(1)).resolveId(`object`)
    }
}
